docs/design/libraries/ComInterfaceGenerator/VTableStubs.md

```diff
@@ -0,0 +1,399 @@
+
+```csharp
+namespace System.Runtime.InteropServices;
+
+public readonly ref struct VirtualMethodTableInfo
+{
+    public VirtualMethodTableInfo(IntPtr thisPointer, ReadOnlySpan<IntPtr> virtualMethodTable)
+    {
+        ThisPointer = thisPointer;
+        VirtualMethodTable = virtualMethodTable;
+    }
+
+    public IntPtr ThisPointer { get; }
+    public ReadOnlySpan<IntPtr> VirtualMethodTable { get; }
+
+    public void Deconstruct(out IntPtr thisPointer, out ReadOnlySpan<IntPtr> virtualMethodTable) // This method allows tuple-style `var (thisPtr, vtable) = virtualMethodTableInfo;` statements from this type.
+    {
+        thisPointer = ThisPointer;
+        virtualMethodTable = VirtualMethodTable;
+    }
+}
+
+public interface IUnmanagedVirtualMethodTableProvider<T> where T : IEquatable<T>
+{
+    VirtualMethodTableInfo GetVirtualMethodTableInfoForKey(T typeKey);
+}
+```
+
+## Required API Shapes
+
+In addition to the provided APIs above, users will be required to add a `readonly static` field or `get`-able property to their user-defined interface type named `TypeKey`. The type of this member will be used as the `T` in `IUnmanagedVirtualMethodTableProvider<T>` and the value will be passed to `GetVirtualMethodTableInfoForKey`. This mechanism is designed to enable each native API platform to provide their own casting key, for example `IID`s in COM, without interfering with each other or requiring using reflection-based types like `System.Type`.
+
+## Example Usage
+
+### Flat function table
+
+In this example, the native API provides a flat table of functions based on the provided version.
+
+```cpp
+// NativeAPI.cpp
+
+struct NativeAPI
+{
+    int(*getVersion)();
+    int(*add)(int x, int y);
+    int(*multiply)(int x, int y);
+};
+
+namespace
+{
+    int getVersion()
+    {
+        return 1;
+    }
+    int add(int x, int y)
+    {
+        return x + y;
+    }
+    int multiply(int x, int y)
+    {
+        return x * y;
+    }
+    const NativeAPI g_nativeAPI = {
+        &getVersion,
+        &add,
+        &multiply
+    };
+}
+
+extern "C" bool GetNativeAPI(int version, NativeAPI const** ppNativeAPI)
+{
+    if (version > getVersion())
+    {
+        *ppNativeAPI = nullptr;
+        return false;
+    }
+    *ppNativeAPI = &g_nativeAPI;
+    return true;
+}
+
+```
+
+```csharp
+// User-written code
+// NativeAPI.cs
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+[assembly:DisableRuntimeMarshalling]
```

Comment:
We should document this if it plan to make it a requirement. If not, I would remove it from the example.